src/main/java/com/mraof/minestuck/block/CassettePlayerBlock.java

```diff
@@ -0,0 +1,129 @@
+import net.minecraft.util.math.BlockRayTraceResult;
+import net.minecraft.world.IBlockReader;
+import net.minecraft.world.IWorld;
+import net.minecraft.world.World;
+
+import javax.annotation.Nullable;
+import java.util.function.Supplier;
+
+
+public class CassettePlayerBlock extends DecorBlock {
+    public static final BooleanProperty HAS_CASSETTE = ModBlockStateProperties.HAS_CASSETTE;
+    public static final BooleanProperty IS_OPEN = ModBlockStateProperties.IS_OPEN;
+    private final Supplier<TileEntityType<?>> tileType;
+
+    public CassettePlayerBlock(Properties properties, CustomVoxelShape shape, Supplier<TileEntityType<?>> tileType) {
+        super(properties, shape);
+        this.setDefaultState(this.stateContainer.getBaseState().with(HAS_CASSETTE, Boolean.valueOf(false)));
+        this.tileType = tileType;
+    }
+
+    @Override
+    @SuppressWarnings("deprecation")
+    public boolean onBlockActivated(BlockState state, World worldIn, BlockPos pos, PlayerEntity player, Hand handIn, BlockRayTraceResult hit) {
+        if (player.isSneaking()) {
+            if (state.get(IS_OPEN)) state = state.with(IS_OPEN, Boolean.valueOf(false));
+            else state = state.with(IS_OPEN, Boolean.valueOf(true));
+            worldIn.setBlockState(pos, state, 2);
+            return true;
+        } else if (state.get(HAS_CASSETTE) && state.get(IS_OPEN)) {
+            this.dropCassette(worldIn, pos);
+            state = state.with(HAS_CASSETTE, Boolean.valueOf(false));
+            worldIn.setBlockState(pos, state, 2);
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    public void insertCassette(IWorld worldIn, BlockPos pos, BlockState state, ItemStack cassetteStack) {
+        TileEntity tileentity = worldIn.getTileEntity(pos);
+        if (tileentity instanceof CassettePlayerTileEntity && state.get(IS_OPEN)) {
+            ((CassettePlayerTileEntity)tileentity).setCassette(cassetteStack.copy());
+            worldIn.setBlockState(pos, state.with(HAS_CASSETTE, Boolean.valueOf(true)), 2);
+        }
+    }
+
+    private void dropCassette(World worldIn, BlockPos pos) {
+        if (!worldIn.isRemote) {
+            TileEntity tileentity = worldIn.getTileEntity(pos);
+            if (tileentity instanceof CassettePlayerTileEntity) {
+                CassettePlayerTileEntity cassetteplayertileentity = (CassettePlayerTileEntity)tileentity;
+                ItemStack itemstack = cassetteplayertileentity.getCassette();
+                if (!itemstack.isEmpty()) {
+                    worldIn.playEvent(1010, pos, 0);
+                    cassetteplayertileentity.clear();
+                    float f = 0.7F;
+                    double d0 = (double)(worldIn.rand.nextFloat() * 0.7F) + (double)0.15F;
+                    double d1 = (double)(worldIn.rand.nextFloat() * 0.7F) + (double)0.060000002F + 0.6D;
+                    double d2 = (double)(worldIn.rand.nextFloat() * 0.7F) + (double)0.15F;
```

Comment:
im kinda afraid of changing the value. Like you may have noticed I used A LOT of mojang work on the jukebox to do jukebox functionality of my block. I don't know why they chose to do that for the random value of the Y axes instead of just writing 0.15F like the two other axes when they want to drop a block, but im afraid of creating a bug they fixed by doing so.

Reply:
That's a typical thing to happen with certain float values when decompiling code. As for accidentaly creating a bug, try considering what the code does to try and judge the possibility of bugs appearing. (It is always better to try to understand and reproduce vanilla code instead of outright copying it)

Reply:
About the variable names, while the random offsets could have a variable name more in tune with their use. (calling it an offset instead of random), and `cassetteplayertileentity` is a bit too long of a variable name (doesn't really need the tileentity part, I'd say),  it's probably good enough to stay as it is.

Reply:
oh no, I didn't write that where it belong. I changed the name there is no problem with that, I meant to write it in the post where you was asking why the random values needed to be so precise.